tests/functional-scenarios.js

```diff
@@ -0,0 +1,297 @@
+      3: [],
+    },
+  };
+
+  let selectRawAuthorData;
+  let selectRawBookData;
+  let selectAuthor;
+  let selectBook;
+  let selectAuthorForBook;
+
+  beforeEach(() => {
+    // The selectors get recreated for each test, to reset their call counts.
+    // The root selector just returns the raw data in the state, while the selector that's built on top
+    // of it will add in some additional fields.
+    selectRawAuthorData = createParameterizedRootSelector(
+      (state, authorId) => state.authorDataById[authorId],
+      {
+        displayName: 'selectRawAuthorData',
+        compareSelectorResults: COMPARISON_PRESETS.SHALLOW_EQUAL,
+        performanceChecksEnabled: true,
+      },
+    );
+    selectRawBookData = createParameterizedRootSelector(
+      (state, bookId) => state.bookDataById[bookId],
+      {
+        displayName: 'selectRawBookData',
+        compareSelectorResults: COMPARISON_PRESETS.SHALLOW_EQUAL,
+        performanceChecksEnabled: true,
+      },
+    );
+
+    // By convention in this test only, root selectors receive an ID directly while intermediate selectors
+    // receive a params object.
+    selectAuthor = createParameterizedSelector(
+      ({ authorId }) => {
+        const rawAuthorData = selectRawAuthorData(authorId);
+        if (!rawAuthorData) {
+          return null;
+        }
+        const name = (rawAuthorData.name || '').trim();
+        return {
+          ...rawAuthorData,
+          authorId,
+          name,
+          nameLowerCase: name.toLowerCase(),
+        };
+      },
+      {
+        displayName: 'selectAuthor',
+        compareSelectorResults: COMPARISON_PRESETS.SHALLOW_EQUAL,
+        performanceChecksEnabled: true,
+      },
+    );
+    selectBook = createParameterizedSelector(
+      ({ bookId }) => {
+        const rawBookData = selectRawBookData(bookId);
+        if (!rawBookData) {
+          return null;
+        }
+        const title = (rawBookData.title || '').trim();
+        return {
+          ...rawBookData,
+          bookId,
+          title,
+          titleLowerCase: title.toLowerCase(),
+        };
+      },
+      {
+        displayName: 'selectBook',
+        compareSelectorResults: COMPARISON_PRESETS.SHALLOW_EQUAL,
+        performanceChecksEnabled: true,
+      },
+    );
+
+    selectAuthorForBook = createParameterizedSelector(
+      (params) => {
+        const book = selectBook(params);
+        if (!book || !book.authorId) {
+          return null;
+        }
+        return selectAuthor({ authorId: book.authorId });
+      },
+      {
+        displayName: 'selectAuthorForBook',
+        performanceChecksEnabled: true,
+      },
+    );
+  });
+
+
+  it('should return author models', () => {
+    const firstAuthor = selectAuthor(initialState, { authorId: 1 });
+    const secondAuthor = selectAuthor(initialState, { authorId: 2 });
+
+    assert.equal(firstAuthor.authorId, 1);
+    assert.equal(firstAuthor, selectAuthor(initialState, { authorId: 1 }));
```

Comment:
opportunity for snapshot tests?

Reply:
:+1: I'll work on this some more this weekend.